File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        dic = {')': '(', '}': '{', ']': '['}
        stack = []

        for i in s:
            if i not in dic.keys():
                stack.append(i)
            else:
                if not stack or stack[-1] != dic[i]: return False
                stack.pop()

        return not stack        

File: 0020-valid-parentheses/0020-valid-parentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        List<Character> close = new ArrayList(Arrays.asList(')', '}', ']'));

        for (char c : s.toCharArray()) {
            if (close.contains(c) && !stack.isEmpty()) {
                if (c == ')' && stack.peek() == '(') {
                    stack.pop();
                } else if (c == '}' && stack.peek() == '{') {
                    stack.pop();
                } else if (c == ']' && stack.peek() == '[') {
                    stack.pop();
                } else {
                    return false;
                }
            } else {
                stack.push(c);
            }
        }

        if (stack.isEmpty()) return true;
        else return false;
    }
}